src/main/java/uk/ac/sanger/sccp/stan/service/NextReplicateServiceImp.java

```diff
@@ -54,66 +54,32 @@ public Map<RepKey, Set<String>> groupBarcodes(Collection<String> barcodes) {
     }
 
     public NextReplicateData toReplicateData(RepKey key, Set<String> barcodes) {
-        Integer repInteger = tissueRepo.findMaxReplicateForDonorIdAndSpatialLocationId(key.getDonorId(), key.getSpatialLocationId());
-        int nextRep = (repInteger==null ? 1 : (repInteger+1));
-        return new NextReplicateData(new ArrayList<>(barcodes), key.getDonorId(), key.getSpatialLocationId(), nextRep);
+        Integer repInteger = tissueRepo.findMaxReplicateForDonorIdAndSpatialLocationId(key.donorId(), key.spatialLocationId());
+        int nextRep = (repInteger == null ? 1 : (repInteger + 1));
+        return new NextReplicateData(new ArrayList<>(barcodes), key.donorId(), key.spatialLocationId(), nextRep);
     }
 
     public Tissue getSingleTissue(Labware lw) {
         Tissue tissue = null;
         for (Slot slot : lw.getSlots()) {
             for (Sample sample : slot.getSamples()) {
-                if (tissue==null) {
+                if (tissue == null) {
                     tissue = sample.getTissue();
                 } else if (!tissue.equals(sample.getTissue())) {
-                    throw new IllegalArgumentException("Labware "+lw.getBarcode()+" contains multiple different tissues.");
+                    throw new IllegalArgumentException("Labware " + lw.getBarcode() + " contains multiple different tissues.");
                 }
             }
         }
-        if (tissue==null) {
-            throw new IllegalArgumentException("Labware "+lw.getBarcode()+" is empty.");
+        if (tissue == null) {
+            throw new IllegalArgumentException("Labware " + lw.getBarcode() + " is empty.");
         }
         return tissue;
     }
 
-    public static class RepKey {
-        public final int donorId;
-        public final int spatialLocationId;
-
-        public RepKey(int donorId, int spatialLocationId) {
-            this.donorId = donorId;
-            this.spatialLocationId = spatialLocationId;
-        }
-
-        public RepKey(Tissue tissue) {
+    record RepKey(int donorId, int spatialLocationId) {
```

Comment:
Just to understand - I hope 'record' is a java21 addition.  What does it mean?

Reply:
Yes. A record is a class where the constructor, the getters, the equals and hashCode and the toString methods can all be generated procedurally so you don't have to write boiler-plate data classes.

Reply:
That sounds cool 👍